iree/hal/api.h

```diff
@@ -966,6 +1014,55 @@ iree_hal_device_allocator(iree_hal_device_t* device);
 IREE_API_EXPORT iree_string_view_t IREE_API_CALL
 iree_hal_device_id(iree_hal_device_t* device);
 
+// Submits one or more batches of work to a device queue.
+//
+// The queue is selected based on the flags set in |command_categories| and the
+// |queue_affinity|. As the number of available queues can vary the
+// |queue_affinity| is used to hash into the available queues for the required
+// categories. For example if 2 queues support transfer commands and the
+// affinity is 5 the resulting queue could  be index hash(5)=1. The affinity can
+// thus be treated as just a way to indicate whether two submissions must be
+// placed on to the same queue. Note that the exact hashing function is
+// implementation dependent.
+//
+// The submission behavior matches Vulkan's vkQueueSubmit, with each batch
+// executing its command buffers in the order they are defined but allowing the
+// command buffers to complete out-of-order. See:
+// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/vkQueueSubmit.html
+IREE_API_EXPORT iree_status_t IREE_API_CALL iree_hal_device_queue_submit(
+    iree_hal_device_t* device, iree_hal_command_category_t command_categories,
+    uint64_t queue_affinity, iree_host_size_t batch_count,
+    const iree_hal_submission_batch_t* batches);
+
+// Blocks the caller until the semaphores reach or exceed the specified payload
+// values or the |deadline_ns| elapses. All semaphores in |semaphore_list| must
+// be created from this device (or be imported into it).
+//
+// |wait_mode| can be used to decide when the wait will proceed; whether *all*
+// semaphores in |semaphore_list| must be signaled or whether *any* (one or
+// more) can be signaled before an early return.
+//
+// Returns success if the wait is successful and semaphores have been signaled
+// satisfying the |wait_mode|.
+//
+// Returns DEADLINE_EXCEEDED if the |deadline_ns| elapses without the
+// |wait_mode| being satisfied. Note that even on success only a subset of the
+// semaphores may have been signaled and each can be queried to see which ones.
+IREE_API_EXPORT iree_status_t IREE_API_CALL
+iree_hal_device_wait_semaphores_with_deadline(
+    iree_hal_device_t* device, iree_hal_wait_mode_t wait_mode,
+    const iree_hal_semaphore_list_t* semaphore_list, iree_time_t deadline_ns);
+
+// Blocks the caller until the semaphores reach or exceed the specified payload
+// values or the |timeout_ns| elapses.
+// A relative-time version of iree_hal_device_wait_semaphores_with_deadline
+// using the relative nanoseconds from the time the call is made.
```

Comment:
It's fine to to not match the IREE timer - since the time is a standard unix wall-clock time it's possible to generate these values in other environments (python, java, etc) in a way that survives queuing/IPC/etc for when it ends up on this side. So if a user wants to provide relative times they use the timeout one and if they want deadlines they get the time based on whatever they want.